async_simple/coro/FutureAwaiter.h

```diff
@@ -31,29 +33,70 @@ namespace coro::detail {
 template <typename T>
 struct FutureAwaiter {
     Future<T> future_;
+    std::shared_ptr<std::pair<Future<T>, std::atomic<bool>>> cancelState_;
+    async_simple::Slot* slot_ = nullptr;
 
     bool await_ready() { return future_.hasResult(); }
 
     template <typename PromiseType>
-    void await_suspend(CoroHandle<PromiseType> continuation) {
+    bool await_suspend(CoroHandle<PromiseType> continuation) {
         static_assert(std::is_base_of_v<LazyPromiseBase, PromiseType>,
                       "FutureAwaiter is only allowed to be called by Lazy");
         Executor* ex = continuation.promise()._executor;
         Executor::Context ctx = Executor::NULLCTX;
         if (ex != nullptr) {
             ctx = ex->checkout();
         }
-        future_.setContinuation([continuation, ex, ctx](Try<T>&& t) mutable {
-            if (ex != nullptr) {
-                ex->checkin(continuation, ctx);
-            } else {
-                continuation.resume();
+        std::shared_ptr<std::pair<Future<T>, std::atomic<bool>>> cancelState;
```

Comment:
This function is too long. Please split it into a "cancellable" version and call it if cancellation is possible and call the current version otherwise.

Reply:
Done